### PR TITLE
[RHCLOUD-17533] Bulk Create

### PR DESCRIPTION
**Pull request by lindgrenj6**
The big one. This reimplements bulk_create on the Go side.

It wasn't too bad really, though I may change the endpoint up in the future to just build an object tree instead of separate arrays, but that would be a breaking change. 

- [x] #73 <- built off this one, so this one needs to go first.
- [x] #100 rebased onto this

\# TODO:
- [x] parse authentications
- [x] authtype validation
- [x] save all sources in transaction, monitor for failure
- [x] ~an ungodly amount of tests~ eventually. if this breaks everything breaks so that'll be the initial tests

### CODE REVIEW

File: model/application_authentication.go
Comment (lpichler) [inline model/application_authentication.go:19]: `gorm: "-"` it doesn't create neither the column ?

Why we need `gorm:"-"` here ? I tough we are using it. Statuslistener's tests  started to fail(strange thing is that it is different error that here in CI) with this change.

Comment (lindgrenj6) [reply model/application_authentication.go:19]: It looks like it's the `AuthenticationUID` that is causing the tests to fail:
```diff
diff --git a/model/application_authentication.go b/model/application_authentication.go
index 50753ac..b979683 100644
--- a/model/application_authentication.go
+++ b/model/application_authentication.go
@@ -26,7 +26,7 @@ type ApplicationAuthentication struct {
 	// TODO: fix correctly PR#40
 	AuthenticationID int64 `json:"authentication_id"`
 
-	AuthenticationUID string `json:"-" gorm:"-"`
+	AuthenticationUID string `json:"-"`
 }
 
 func (aa *ApplicationAuthentication) ToEvent() interface{} {
```

Which is a little unfortunate - the reason I removed the VaultPath was because that column isn't in the DB yet so gorm fails when that column isn't there (e.g. outside the tests). 

The `AuthenticationUID` piece may be something we have to play with a bit more - we could just store it (in the migration on the rails side when we do the data migration from postgres -> vault) but then we'd have duplicated data (VaultPath + UID). 

I'm thinking we'll just store it in one place (Path), then transform it into the ID field that way the frontend doesn't have to change anything (the authentication_id will just return the UUID instead of the string representation of the pkey). We'll also have to remove the foreign key there. All sorts of fun. 

---

All in all I removed the `gorm:"-"` from the AuthenticationUID for now so the tests pass. We'll worry about making the application authentication "correct" when we do the actual migration to vault.

**Comment by lindgrenj6** [comment]
@lpichler @MikelAlejoBR :eye: plox and thanks!

File: service/bulk_create.go
Comment (MikelAlejoBR) [inline service/bulk_create.go:266]: I think this should be `auth.ResourceId`

File: dao/application_type_dao.go
Comment (MikelAlejoBR) [inline dao/application_type_dao.go]: I think a `.Debug` statement is missing here

File: dao/source_type_dao.go
Comment (MikelAlejoBR) [inline dao/source_type_dao.go]: Missing a `.Debug` statement

File: service/bulk_create.go
Comment (lindgrenj6) [reply service/bulk_create.go:266]: Actually this is a fun little piece of functionality that I'm not sure anyone uses - you can attach an authentication to an existing application/endpoint/source. In the bulk create payload you'd just use `{resource_name: "42", resource_type: "Application"}` in the authentication part of the payload. 

I'm not sure anyone actually uses it - but thats what this is for. Checking to see if an int was passed in or not.

Comment (MikelAlejoBR) [reply service/bulk_create.go:266]: Aaah got it. Thanks for the explanation!

**Comment by lpichler** [comment]
I did:

```
POST http://localhost:{{port}}/api/sources/v3.1/bulk_create
Content-Type: application/json; charset=UTF-8
x-rh-identity: {{x-rh-identity}}

{
  "sources": [
    {
      "name": "my amazon sourceGO",
      "source_type_name": "amazon"
    }
  ],
  "applications": [
    {
      "source_name": "my amazon sourceGO",
      "application_type_name": "cost-management"
    },
    {
      "source_name": "my amazon sourceGO",
      "application_type_name": "cloud-meter"
    }
  ],
  "authentications": [
    {
      "resource_type": "source",
      "resource_name": "my amazon sourceGO",
      "username": "aws::iam::arn::i_cant_remember_the_format",
      "authtype": "arn"
    },
    {
      "resource_type": "application",
      "resource_name": "cost-management",
      "username": "myUser",
      "password": "myPass"
    },
    {
      "resource_type": "application",
      "resource_name": "cloud-meter",
      "username": "cloudUser",
      "password": "cloudPass"
    }
  ]
}
```
response
```

HTTP/1.1 500 Internal Server Error
Content-Type: application/json; charset=UTF-8
X-Request-Duration-Ms: 28
Date: Fri, 25 Mar 2022 13:51:29 GMT
Content-Length: 202

{
  "errors": [
    {
      "detail": "Internal Server Error: ERROR: insert or update on table \"application_authentications\" violates foreign key constraint \"fk_rails_d709bbbff3\" (SQLSTATE 23503)",
      "status": "500"
    }
  ]
}
```

```
INSERT INTO \"application_authentications\" (\"paused_at\",\"created_at\",\"updated_at\",\"tenant_id\",\"application_id\",\"authentication_id\",\"authentication_uid\") VALUES ('0000-00-00 00:00:00','2022-03-25 14:51:29.833','2022-03-25 14:51:29.833',1,7,0,'3a3f5926-9eff-4fb6-ac38-1b58169d22bb'),('0000-00-00 00:00:00','2022-03-25 14:51:29.833','2022-03-25 14:51:29.833',1,8,0,'7d13a026-aa8c-440a-b48b-ed21efc18c9d') RETURNING \"id\""
```

we need `authentication_id`.
Are we going to leave it as it is until we will remove this foreign key ?

**Comment by lindgrenj6** [comment]
Yeah - basically we're going to need to leave it be until we have the DB DAO merged. It should "just work" with the DB DAO (after adding the field setting) since that key won't be violated anymore. I may just wait until the postgres backed DB DAO is merged to make sure and update this. It'll just be setting that key and then inserting it separately.

**Review by lpichler — APPROVED** [review]

File: model/application_authentication.go
Comment (lpichler) [reply model/application_authentication.go:19]: yea I think we can do it with just storing in Path.
thanks!

**Comment by lindgrenj6** [comment]
Yeah - basically we're going to need to leave it be until we have the DB
DAO merged. It should "just work" with the DB DAO since that key won't be
violated.
--
Jacob Lindgren

Senior Software Engineer - Insights Platform
Red Hat
***@***.***     M: 402-841-6173

**Pull request merged**
